lib/ddtrace/pipeline.rb

```diff
@@ -0,0 +1,42 @@
+module Datadog
+  # Pipeline
+  module Pipeline
+    require_relative 'pipeline/span_filter'
+    require_relative 'pipeline/span_processor'
+
+    @mutex = Mutex.new
+    @processors = []
+
+    def self.before_flush(*processors, &processor_block)
+      processors = [processor_block] if processors.empty?
+
+      @mutex.synchronize do
+        @processors.concat(processors)
+      end
+    end
+
+    def self.process!(traces)
+      @mutex.synchronize do
+        traces
+          .map(&method(:apply_processors!))
+          .select(&:any?)
+      end
+    end
+
+    def self.processors=(value)
+      @processors = value
+    end
+
+    def self.apply_processors!(trace)
+      result = @processors.inject(trace) do |current_trace, processor|
+        processor.call(current_trace)
+      end
+
+      result || []
+    rescue
+      []
```

Comment:
let's add a `log.debug` here saying that a trace will be dropped because of an exception in the processor (we may show also the error message I think)